src/main/kotlin/org/rust/ide/inspections/RsLivenessInspection.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.psi.PsiElement
+import org.rust.ide.injected.isDoctestInjection
+import org.rust.ide.inspections.fixes.RenameFix
+import org.rust.ide.utils.isCfgUnknown
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.descendantsWithMacrosOfType
+import org.rust.lang.core.psi.ext.expansion
+import org.rust.lang.core.types.DeclarationKind
+import org.rust.lang.core.types.DeclarationKind.Parameter
+import org.rust.lang.core.types.DeclarationKind.Variable
+import org.rust.lang.core.types.liveness
+
+class RsLivenessInspection : RsLintInspection() {
+
+    override fun getLint(element: PsiElement): RsLint? =
+        RsLint.UnusedVariables
+
+    override fun buildVisitor(holder: RsProblemsHolder, isOnTheFly: Boolean) =
+        object : RsVisitor() {
+            override fun visitFunction(func: RsFunction) {
+                // Disable inside doc tests
+                if (func.isDoctestInjection) return
+
+                // Don't analyze functions with unresolved macro calls
+                if (func.descendantsWithMacrosOfType<RsMacroCall>().any { it.expansion == null }) return
```

Comment:
@vlad20012 looks like you had some questions about this

Reply:
I rethought it. Now I think everything is fine here